File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int m = s.size();
        int n = t.size();

        vector<int> dp(n + 1, 0);
        int MOD = (int)1e9 + 7; // to avoid int overflow

        for (int i = 1; i <= m; i++) {
            int last = 1;

            for (int j = 1; j <= n; j++) {
                int cur = dp[j];
                int take = 0, notTake = 0;
                
                if (s[i - 1] == t[j - 1]) {
                    take = last;
                }
                notTake = cur;
                dp[j] = (take + notTake) % MOD;
                last = cur;
            }
        }
        return dp[n];
    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.ts
function numDistinct(s: string, t: string): number {
    const MOD = 1e9 + 7;
    const n = s.length,
        m = t.length;
    const dp: number[] = Array(m + 1).fill(0);
    dp[0] = 1;

    for (let i = 1; i <= n; i++) {
        let prev = dp[0];
        for (let j = 1; j <= m; j++) {
            const cur = dp[j];
            if (s[i - 1] === t[j - 1]) {
                dp[j] = (dp[j] + prev) % MOD;
            }
            prev = cur;
        }
    }
    return dp[m];
}